flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java

```diff
@@ -0,0 +1,259 @@
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.AlreadyExistsException;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.InvalidOperationException;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A catalog that persists all Flink streaming and batch metadata by using Hive metastore as a persistent storage.
+ */
+public class GenericHiveMetastoreCatalog implements ReadableWritableCatalog {
+	private static final Logger LOG = LoggerFactory.getLogger(GenericHiveMetastoreCatalog.class);
+
+	public static final String DEFAULT_DB = "default";
+
+	private final String catalogName;
+	private final HiveConf hiveConf;
+
+	private String currentDatabase = DEFAULT_DB;
+	private IMetaStoreClient client;
+
+	public GenericHiveMetastoreCatalog(String catalogName, String hivemetastoreURI) {
+		this(catalogName, getHiveConf(hivemetastoreURI));
+	}
+
+	public GenericHiveMetastoreCatalog(String catalogName, HiveConf hiveConf) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(catalogName), "catalogName cannot be null or empty");
+		this.catalogName = catalogName;
+
+		this.hiveConf = checkNotNull(hiveConf, "hiveConf cannot be null");
+		LOG.info("Created GenericHiveMetastoreCatalog '{}'", catalogName);
+	}
+
+	private static HiveConf getHiveConf(String hiveMetastoreURI) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(hiveMetastoreURI), "hiveMetastoreURI cannot be null or empty");
+
+		HiveConf hiveConf = new HiveConf();
+		hiveConf.setVar(HiveConf.ConfVars.METASTOREURIS, hiveMetastoreURI);
+		return hiveConf;
+	}
+
+	private static IMetaStoreClient getmetastoreClient(HiveConf hiveConf) {
+		try {
+			return RetryingMetaStoreClient.getProxy(
+				hiveConf,
+				null,
+				null,
+				HiveMetaStoreClient.class.getName(),
+				true);
+		} catch (MetaException e) {
+			throw new CatalogException("Failed to create Hive metastore client", e);
+		}
+	}
+
+	@Override
+	public void open() throws CatalogException {
+		if (client == null) {
+			client = getmetastoreClient(hiveConf);
+			LOG.info("Connected to Hive metastore");
+		}
+	}
+
+	@Override
+	public void close() throws CatalogException {
+		if (client != null) {
+			client.close();
+			client = null;
+			LOG.info("Close connection to Hive metastore");
+		}
+	}
+
+	// ------ databases ------
+
+	@Override
+	public String getCurrentDatabase() throws CatalogException {
```

Comment:
Should we change CatalogException to checked Exception ? Currently CatalogException extends RuntimeException.

Reply:
There was a discussion for it in previous PR https://github.com/apache/flink/pull/8007 
AFAICT, getCurrentDatabase() is so dead simple that it probably would never throw any exception and doesn't need to have this signature, we can revisit this later on.